cscs-checks/prgenv/affinity_check.py

```diff
@@ -21,166 +35,523 @@ def __init__(self, variant):
-            re_aff_cores, self.stdout, 'cpus', parse_cpus)
-        ref_key = 'ref_' + self.current_partition.fullname
-        self.ref_cores = sn.extractall(
-            re_aff_cores, self.cases[self.variant][ref_key],
-            'cpus', parse_cpus)
+        The cpu set can be extracted by matching core, numa domain or socket.
+        This is controlled by the `by` argument.
+        '''
+        _map = {
+            'core': 3,
+            'socket': 2,
+            'numa': 1,
+        }
+
+        if (cpuid < 0) or (cpuid >= self.num_cpus):
+            raise TaskExit(f'a cpuid with value {cpuid} is invalid')
+
+        if by is None:
+            raise TaskExit('must specify the sibiling level')
+        else:
+            if by not in self._lscpu[0]:
+                raise TaskExit('invalid sibiling level')
+
+        sibiling_id = list(
+            filter(lambda x: int(x['cpu']) == cpuid, self._lscpu)
+        )[0][by]
+
+        return set(
+            map(
+                lambda y: int(y['cpu']),
+                filter(
+                    lambda x: x[by] == sibiling_id,
+                    self._lscpu
+                )
+            )
+        )
+
+    @sn.sanity_function
+    def assert_consumed_cpu_set(self):
+        '''Check that all the resources have been consumed.
+
+        Tests derived from this class must implement a hook that consumes
+        the cpu set as the results from the affinity tool are processed.
+        '''
+        return sn.assert_eq(self.cpu_set, set())
+
+    @rfm.run_after('run')
+    def parse_output(self):
+        '''Extract the data from the affinity tool.'''
+        re_aff_cpus = r'CPU affinity: \[\s+(?P<cpus>[\d+\s+]+)\]'
         re_aff_thrds = r'^Tag:[^\n\r]*Thread:\s+(?P<thread>\d+)'
-        self.aff_thrds = sn.extractall(re_aff_thrds, self.stdout, 'thread',
-                                       int)
-        self.ref_thrds = sn.extractall(
-            re_aff_thrds, self.cases[self.variant][ref_key],
-            'thread', int)
         re_aff_ranks = r'^Tag:[^\n\r]*Rank:\s+(?P<rank>\d+)[\s+\S+]'
-        self.aff_ranks = sn.extractall(re_aff_ranks, self.stdout, 'rank', int)
-        self.ref_ranks = sn.extractall(
-            re_aff_ranks, self.cases[self.variant][ref_key],
-            'rank', int)
-
-        # Ranks and threads can be extracted into lists in order to compare
-        # them since the affinity programm prints them in ascending order.
-        self.sanity_patterns = sn.all([
-            sn.assert_eq(self.aff_thrds, self.ref_thrds),
-            sn.assert_eq(self.aff_ranks, self.ref_ranks),
-            sn.assert_eq(sn.sorted(self.aff_cores), sn.sorted(self.ref_cores))
-        ])
+
+        def parse_cpus(x):
+            return sorted([int(xi) for xi in x.split()])
+
+        with osext.change_dir(self.stagedir):
+            self.aff_cpus = sn.extractall(
+                re_aff_cpus, self.stdout, 'cpus', parse_cpus
+            ).evaluate()
```

Comment:
Why do you need `evaluate()` here and not in the the rest?

Reply:
I just copied those over from the previous test, but they're not used for these new tests. That's why I didn't notice that they had that `evaluate()` missing.

I'm taking `aff_thrds` and `aff_ranks` out.